update/v1alpha1/types_machineconfigpool_progress_insight.go

```diff
@@ -0,0 +1,173 @@
+	//
+	// +Note: By OpenShift API conventions, in isolation this should be a specialized reference that refers just to
+	// +resource name (because the rest is implied by status insight type). However, because we use resource references in
+	// +many places and this API is intended to be consumed by clients, not produced, consistency seems to be more valuable
+	// +than type safety for producers.
+	// +required
+	// +kubebuilder:validation:XValidation:rule="self.group == 'machineconfiguration.openshift.io' && self.resource == 'machineconfigpools'",message="resource must be a machineconfigpools.machineconfiguration.openshift.io resource"
+	Resource ResourceRef `json:"resource"`
+
+	// scopeType describes whether the pool is a control plane or a worker pool
+	// +required
+	Scope ScopeType `json:"scopeType"`
+
+	// assessment is the assessment of the machine config pool update process. Valid values are: Pending, Completed, Degraded, Excluded, Progressing
+	// +required
+	Assessment PoolAssessment `json:"assessment"`
+
+	// completion is a percentage of the update completion (0-100)
+	// +required
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=100
+	Completion int32 `json:"completion"`
+
+	// summaries is a list of counts of nodes matching certain criteria (e.g. updated, degraded, etc.). Maximum 16 items can be listed.
+	// +listType=map
+	// +listMapKey=type
+	// +patchStrategy=merge
+	// +patchMergeKey=type
+	// +optional
+	// +kubebuilder:validation:MaxItems=16
+	Summaries []NodeSummary `json:"summaries,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
+}
+
+// PoolAssessment is the assessment of the node pool update process
+// +kubebuilder:validation:Enum=Pending;Completed;Degraded;Excluded;Progressing
+type PoolAssessment string
+
+const (
+	// Pending means the nodes in the pool will be updated but none have even started yet
+	PoolPending PoolAssessment = "Pending"
+	// Completed means all nodes in the pool have been updated
+	PoolCompleted PoolAssessment = "Completed"
+	// Degraded means the process of updating the pool suffers from an observed problem
+	PoolDegraded PoolAssessment = "Degraded"
+	// Excluded means some (or all) nodes in the pool would be normally updated but a configuration (such as paused MCP)
+	// prevents that from happening
+	PoolExcluded PoolAssessment = "Excluded"
+	// Progressing means the nodes in the pool are being updated and no problems or slowness are detected
+	PoolProgressing PoolAssessment = "Progressing"
+)
+
+// NodeSummary is a count of nodes matching certain criteria (e.g. updated, degraded, etc.)
+type NodeSummary struct {
+	// type is the type of the summary. Valid values are: Total, Available, Progressing, Outdated, Draining, Excluded, Degraded
+	// The summaries are not exclusive, a single node may be counted in multiple summaries.
+	// +required
+	Type NodeSummaryType `json:"type"`
+
+	// count is the number of nodes matching the criteria, between 0 and 1024
```

Comment:
I suspect we support more than 1024 nodes, perhaps make the limit 5000? Probably want to check with perf and scale, but I believe our limit is well below 5000

Reply:
https://docs.redhat.com/en/documentation/openshift_container_platform/4.18/html/scalability_and_performance/planning-your-environment-according-to-object-maximums#cluster-maximums-major-releases_object-limits

says 2k

Reply:
Shall we set the limit to 2k then?

Reply:
Why not